src/autorest.bicep/src/schema-generator.ts

```diff
@@ -0,0 +1,729 @@
+      }
+
+      const subTypes = flattenDiscriminatorSubTypes(value);
+      for (const subTypeKey in subTypes) {
+        output[subTypeKey] = subTypes[subTypeKey];
+      }
+    }
+
+    return output;
+  }
+
+  function* getDiscriminatedSubTypes(putSchema: ObjectSchema | undefined) {
+    const putSubTypes = flattenDiscriminatorSubTypes(putSchema);
+
+    for (const subTypeName of uniq([...keys(putSubTypes)])) {
+      yield {
+        subTypeName,
+        putSubType: putSubTypes[subTypeName],
+      };
+    }
+  }
+
+  function parseType(putSchema: Schema | undefined, supportsExpression: boolean) {
+    const schema = parseTypeInternal(putSchema);
+
+    if (schema === undefined ||
+        isEmpty(schema) ||
+        (schema.type === 'string' && !schema.enum && !schema.pattern) ||
+        !supportsExpression) {
+      return schema;
+    }
+
+    return addExpressionOneOf(schema);
+  }
+
+  function addExpressionOneOf(schema: JSONSchema4) {
+    return {
+      oneOf: [
+        schema,
+        {
+          '$ref': 'https://schema.management.azure.com/schemas/common/definitions.json#/definitions/expression',
+        }
+      ]
+    };
+  }
+
+  function parseTypeInternal(putSchema: Schema | undefined): JSONSchema4 | undefined {
+    // A schema that matches a JSON object with specific properties, such as
+    // { "name": { "type": "string" }, "age": { "type": "number" } }
+    if (putSchema instanceof ObjectSchema) {
+      return parseObjectType(putSchema as ObjectSchema);
+    }
+
+    // A schema that matches a "dictionary" JSON object, such as
+    // { "additionalProperties": { "type": "string" } }
+    if (putSchema instanceof DictionarySchema) {
+      return parseDictionaryType(putSchema as DictionarySchema);
+    }
+
+    // A schema that matches a single value from a given set of values, such as
+    // { "enum": [ "a", "b" ] }
+    if (putSchema instanceof ChoiceSchema) {
+      return parseEnumType(putSchema as ChoiceSchema);
+    }
+    if (putSchema instanceof SealedChoiceSchema) {
+      return parseEnumType(putSchema as SealedChoiceSchema);
+    }
+    if (putSchema instanceof ConstantSchema) {
+      return parseConstant(putSchema as ConstantSchema);
+    }
+
+    // A schema that matches an array of values, such as
+    // { "items": { "type": "number" } }
+    if (putSchema instanceof ArraySchema) {
+      return parseArrayType(putSchema as ArraySchema);
+    }
+
+    // A schema that matches simple values (or that is serialized to simple values), such
+    // as { "type": "number" } or { "type": "string", "format": "base64url" }
+    if (putSchema instanceof PrimitiveSchema || (putSchema instanceof ByteArraySchema && putSchema.format !== 'byte')) {
+      return parsePrimaryType(putSchema as PrimitiveSchema);
+    }
+
+    if (putSchema instanceof AnyObjectSchema) {
+      return {
+        type: 'object',
+        properties: {},
+      };
```

Comment:
Should `properties` just be omitted here?

Reply:
Good point. This actually makes a big difference to Export - it's the difference between "allow any properties" and "remove all properties". Updated.